crates/wasmtime/src/memory.rs

```diff
@@ -305,96 +313,67 @@ impl Memory {
+        buffer: &mut [u8],
+    ) -> Result<(), MemoryAccessError> {
+        let store = store.as_context();
+        let slice = self
+            .data(&store)
+            .get(offset..)
+            .and_then(|s| s.get(..buffer.len()))
+            .ok_or(MemoryAccessError { _private: () })?;
+        buffer.copy_from_slice(slice);
+        Ok(())
     }
 
     /// Safely writes contents of a buffer to this memory at the given offset.
     ///
     /// If the offset + buffer length exceed current memory capacity, then none
     /// of the buffer is written to memory and a [`MemoryAccessError`] is
     /// returned.
-    pub fn write(&self, offset: usize, buffer: &[u8]) -> Result<(), MemoryAccessError> {
-        unsafe {
-            self.data_unchecked_mut()
-                .get_mut(offset..)
-                .and_then(|s| s.get_mut(..buffer.len()))
-                .ok_or(MemoryAccessError { _private: () })?
-                .copy_from_slice(buffer);
-            Ok(())
-        }
+    pub fn write(
+        &self,
+        mut store: impl AsContextMut,
+        offset: usize,
+        buffer: &[u8],
+    ) -> Result<(), MemoryAccessError> {
+        let mut context = store.as_context_mut();
+        self.data_mut(&mut context)
+            .get_mut(offset..)
+            .and_then(|s| s.get_mut(..buffer.len()))
+            .ok_or(MemoryAccessError { _private: () })?
+            .copy_from_slice(buffer);
+        Ok(())
     }
 
     /// Returns this memory as a slice view that can be read natively in Rust.
-    ///
-    /// # Safety
-    ///
-    /// This is an unsafe operation because there is no guarantee that the
-    /// following operations do not happen concurrently while the slice is in
-    /// use:
-    ///
-    /// * Data could be modified by calling into a wasm module.
-    /// * Memory could be relocated through growth by calling into a wasm
-    ///   module.
-    /// * When threads are supported, non-atomic reads will race with other
-    ///   writes.
-    ///
-    /// Extreme care need be taken when the data of a `Memory` is read. The
-    /// above invariants all need to be upheld at a bare minimum, and in
-    /// general you'll need to ensure that while you're looking at slice you're
-    /// the only one who can possibly look at the slice and read/write it.
-    ///
-    /// Be sure to keep in mind that `Memory` is reference counted, meaning
-    /// that there may be other users of this `Memory` instance elsewhere in
-    /// your program. Additionally `Memory` can be shared and used in any number
-    /// of wasm instances, so calling any wasm code should be considered
-    /// dangerous while you're holding a slice of memory.
-    ///
-    /// For more information and examples see the documentation on the
-    /// [`Memory`] type.
-    pub unsafe fn data_unchecked(&self) -> &[u8] {
-        self.data_unchecked_mut()
+    pub fn data<'a, T: 'a>(&self, store: impl Into<StoreContext<'a, T>>) -> &'a [u8] {
```

Comment:
Note that this doesn't take `impl AsContext` so we can tie to the precise lifetime.